### ISSUE TITLE
Race condition errors in Task have unhelpful error messages

### ISSUE THREAD

**Issue by robotlolita** [origamitower/folktale#153]
The error messages for race errors in Task is unhelpful:

```text
_deferred.js:49 Uncaught Error: Only pending deferreds can be resolved, this deferred is already resolved.
```

### Steps to reproduce

This happens after trying to settle a Task's deferred that has already settled. It errors because this is most likely an error in the Task's code.

```js
const { task } = require('folktale/concurrency/task');
const ohno = task(resolver => {
  resolver.resolve(1);
  resolver.resolve(2); // <- an error
});

ohno.run();
// => Uncaught Error: Only pending deferreds can be resolved, this deferred is already resolved.
```

### Additional information

While this is how it should work (ignoring those errors may lead to even more confusing problems elsewhere!), the error message is not very helpful. To begin with, the stack trace is from an internal Deferred type, and it's hard to see what it's related to.

The error message should try to include where people should be looking for the error. We could also consider providing safe versions of resolve/reject so people can be intentional about how they expect that call to go, something like `resolveIfNotSettled`.

**Comment by rostero1** [comment]
Adding the demo that I was struggling with for reference:  https://www.webpackbin.com/bins/-KsP9EQYRTO_i-SLQ_xM

**Comment by rostero1** [comment]
One thing possibly worth noting is that `resolver.isCancelled` always prints `false` on line 8 even though I get `Uncaught Error: Only pending deferreds can be resolved, this deferred is already resolved`.

**Comment by joaomilho** [comment]
It also happens when I have 2 deferred tasks and it can't release some resource. Ex:

```
const { task, waitAny } = require('folktale/concurrency/task')
const fs = require('fs')

const writeFile = (path, content, encoding = 'utf8') =>
  task(resolver => {
    fs.writeFile(path, content, encoding, err => {
      if (err) resolver.reject(err)

      resolver.resolve(`Written ${path}`)
    })
    resolver.cleanup(() => {})
    resolver.onCancelled(() => {
      console.log('cancelled...')
    })
  })

waitAny([
  writeFile('./a.txt', 'aha', 'utf8'),
  writeFile('./b.txt', 'aha', 'utf8'),
])
  .map(console.log)
  .run()
```

**Comment by joaomilho** [comment]
Also, I can sorta solve it by doing:
```
const writeFile = (path, content, encoding = 'utf8') =>
  task(resolver => {
    let cancelled = false
    fs.writeFile(path, content, encoding, err => {
      if (err) resolver.reject(err)

      if (cancelled) return
      console.log(`Written ${path}`)
      resolver.resolve(`Written ${path}`)
    })
    resolver.cleanup(() => {})
    resolver.onCancelled(() => {
      cancelled = true
      console.log(`Cancelled ${path}`)
    })
  })
```

**Comment by robotlolita** [comment]
@rostero1 mmh, yeah, `.isCancelled` should work. I'll investigate this further through this week.

**Comment by robotlolita** [comment]
Superseded by #178

**Comment by cdoublev** [comment]
This error also happens when more than one parallel tasks are rejected. Please forgive/correct me if I'm wrong in my following thinking.

Current (and expected) behavior:

```js
const { of, rejected, task, waitAll } = require('folktale/concurrency/task')

const delay = (result, isError = false) => task(resolver =>
    setTimeout(() => isError ? resolver.reject(result) : resolver.resolve(result), 0))

waitAll([delay('a'), delay('b', true)])
    .orElse(error => rejected(console.log('Error: ', error)))
    .run() // => Error: b
```

But with multiple rejections:

```js
waitAll([delay('a'), delay('b', true), delay('c', true)])
    .orElse(e => rejected(console.log('Error: ', e)))
    .run() // => Error: b
    // Uncaught Error: Only pending deferreds can be resolved, this deferred is already resolved.

delay('a').and(delay('b', true).and(delay('c', true)))
    .orElse(error => rejected(console.log('Error: ', error)))
    .run() // => Error: b
    // Uncaught Error: Only pending deferreds can be resolved, this deferred is already resolved.

waitAll([of('a'), rejected('b'), rejected('c')])
    .orElse(e => Task.rejected(console.log('Error: ', e)))
    .run() // => Error: b
```

Is this the exepected behavior?

Yet I can solve it with:

```js
 of(res1 => res2 => res3 => console.log('Results: ', [res1, res2, res3]))
    .apply(delay('a'))
    .apply(delay('b', true))
    .apply(delay('c', true))
    .orElse(error => rejected(console.log('Error: ', error)))
    .run() // => Error: b
```

I have this issue in real code by `fs.read`ing multiple directories. I was writing tests (w/ Mocha) for unfound paths provided by the user.

```js
const assert = require('assert')

it ('should reject task if paths are not found', () => 
    assert.rejects(getFilesNames(paths).run().promise()))

**Comment by robotlolita** [comment]
@cdoublev it is the expected behaviour, but things are a bit complicated here when working with tasks.

Let's say you have the following piece of code:

```js
waitAll([delay('a'), delay('b', true), delay('c', true)])
    .orElse(e => rejected(console.log('Error: ', e)))
    .run() // => Error: b
```

It'll first run `delay('a')`, which resolves correctly.

It'll then move on to run `delay('b', true)`, but this one fails, and rejects the task. This rejection causes the entire `waitAll` task to be cancelled. Meaning that tasks you're waiting on that haven't settled yet will be cancelled, and the entire `waitAll` task will be rejected with the same error `delay('b', true)` returned.

Now, this cancellation is synchronous, so that happens before `delay('c', true)` has a chance to run its callback (because that has been scheduled for the next turn of the event loop, and we haven't yielded control back to Node yet). So when the callback for `delay('c', true)` finally runs, the task has already been settled (as cancelled), and when you try to resolve or reject it, it'll tell you that you can't do that--someone has already settled the task, and you've got a race condition at some point.

Now, the message says "Only pending deferreds can be rejected, this deferred is already rejected.". That's a bug. Which should be fixed by #214.

But going back to the problem here. Because we don't want things to fail silently, this causes an error to be thrown. On one hand, this allows you to know that there's a problem. On the other hand, the messages are totally not helpful (yet), so I've been wondering if this is actually helping or not...

To prevent this from happening you need to explicitly check for cancellation before resolving/rejecting a task that may be cancelled, so:

```js
const delay = (result, isError = false) => task(resolver =>
    setTimeout(() => {
        if (resolver.isCancelled) return; // this task has been cancelled somewhere
        isError ? resolver.reject(result) : resolver.resolve(result)
    }, 0))
```

**Comment by cdoublev** [comment]
Thanks for the feed back @robotlolita, very appreciated.

I also figured it out what's happening by `apply`ing multiple rejected tasks, and why it doesn't throw errors as with `wait`ing`All` of them, or more exactly, *all of them or the first failing, **nothing more***. The last part was somewhat missing in my understanding of the [documentation of `waitAll`](https://folktale.origamitower.com/api/v2.3.0/en/folktale.concurrency.task.wait-all.waitall.html):

> Constructs a new task that waits all given tasks resolve. The result of the new task is an array with the results of the input tasks, if they all succeed, or the first failure if they fail.

But I voluntary pulled out a detail in my previous comment. I'm facing this error by using `Task.fromNodeBack` in my real code. There is a lot using this helper. I almost wrote an entire library for node `fs` using them. It would be a little painfull to rewrite it with your suggested fix. Do you think this fix would be worth to apply in [`Task.fromNodeBack`](https://github.com/origamitower/folktale/blob/master/packages/base/source/conversions/nodeback-to-task.js#L23)?

**Comment by robotlolita** [comment]
@cdoublev ah, yes, that fix should be applied to Task.fromNodeback too. All of the Folktale task primitives should be dealing with cancellation. Can you open a separate issue about this?

**Comment by cdoublev** [comment]
Correct me if I'm wrong: this doesn't concern `fromPromisedToTask` as there is currently no way to cancel a promise, but only `fromNodeback`. I hope that I don't narrowed down the issue in my description.
